src/server/journal/streamer.cc

```diff
@@ -6,27 +6,59 @@
 
 #include <absl/functional/bind_front.h>
 
+#include <chrono>
+#include <system_error>
+
 #include "base/logging.h"
 
 namespace dfly {
 using namespace util;
 
-void JournalStreamer::Start(io::Sink* dest) {
+void JournalStreamer::PeriodicPing::MaybePing(bool allow_await) {
+  const auto now = std::chrono::system_clock::now();
+  const auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - start_time_);
+  if (elapsed > kLimit) {
+    base::IoBuf tmp;
+    io::BufSink sink(&tmp);
+    JournalWriter writer(&sink);
+    journal::Entry entry(0, journal::Op::PING, 0, 0, nullopt, {},
+                         streamer_->journal_->PostIncrLsn());
+    writer.Write(entry);
+
+    journal::JournalItem dummy;
+    streamer_->Write(io::Buffer(io::View(tmp.InputBuffer())));
+    streamer_->NotifyWritten(allow_await);
+    Start();
+  }
+}
+
+void JournalStreamer::PeriodicPing::Start() {
+  start_time_ = std::chrono::system_clock::now();
+}
+
+void JournalStreamer::Start(io::Sink* dest, bool with_pings) {
   using namespace journal;
+  // periodic_ping_ = PeriodicPing(this);
+  periodic_ping_.Start();
   write_fb_ = fb2::Fiber("journal_stream", &JournalStreamer::WriterFb, this, dest);
-  journal_cb_id_ = journal_->RegisterOnChange([this](const JournalItem& item, bool allow_await) {
-    if (!ShouldWrite(item)) {
-      return;
-    }
+  journal_cb_id_ =
+      journal_->RegisterOnChange([this, with_pings](const JournalItem& item, bool allow_await) {
+        if (!ShouldWrite(item)) {
+          return;
+        }
 
-    if (item.opcode == Op::NOOP) {
-      // No record to write, just await if data was written so consumer will read the data.
-      return AwaitIfWritten();
-    }
+        if (item.opcode == Op::NOOP) {
+          // No record to write, just await if data was written so consumer will read the data.
+          return AwaitIfWritten();
+        }
+
+        Write(io::Buffer(item.data));
+        NotifyWritten(allow_await);
 
-    Write(io::Buffer(item.data));
-    NotifyWritten(allow_await);
-  });
+        if (with_pings) {
+          periodic_ping_.MaybePing(allow_await);
```

Comment:
So I dont this flow is working if we have few replicas.
this class is created per replica but here you increase the lsn and send to only single replica the journal change
To make this flow work you should write this code in journal slice so that the ping will be sent to all registered callbacks

Reply:
I am aware but what you suggest would also be problematic and a bug. We use `Journal::RegisterOnChange` in other places and we do not only register callbacks associated with `StableSync`. We also for example register a callback during snapshot (see `snapshot.cc:65`). So if we were to apply the `ping` in all callbacks we would send multiple pings for execution paths that are completely irrelevant. For stable sync, we register `one` callback per flow and each of the flows should periodically ping its lsn.

Reply:
how was this resolved?

Reply:
I am fairly certain that what @adiholden suggests here is a bug. We should `NOT` send a ping for every registered callback because some of the registered callbacks are completely irrelevant with `StableSync`. An example of that is snapshot, that does call `RegisterOnChange` and most certainty you wouldn't want to ping there.

Each shard on master has a local JournalSlice. Even if we have less shards available on the replica, we still have master number of flows. So for example:

```         
master(3 shards)  [shard 1] [shard 2] [shard 3]
                   |       /            |
replica(2 shards)  [shard 1]     [shard 2]
```
Here we have 3 flows each having their local lsn's retrieved from the thread/shard local variable JournalSlice. However, on the replica side, we track LSN's via a local variable of the flow. So, when shard2 sends its lsn to replica shard 1, it does not overwrite a thread local on the replica side, but a member variable that tracks the lsn's for that flow and therefore we should be safe.

Reply:
@kostasrim you resolved this now by not sending the LSN but a local variable you keep track of in the streamer class - you named it total_records_.
With this change we might miss something when trying to understand where is our bug in replica takeover
We conclude replica takeover when the LSN for replica is the same as master
If we missed sending some data from master to replica but it did get to the journal we will not see it. If we sent wrong lsn value when full sync finished we will not see this.
Also regarding my suggestion above it is not a bug, the only think is that we will write a ping to journal when sometimes it is not needed i.e when snapshotting from SAVE command

Reply:
> With this change we might miss something when trying to understand where is our bug in replica takeover
We conclude replica takeover when the LSN for replica is the same as master
If we missed sending some data from master to replica but it did get to the journal we will not see it. If we sent wrong lsn value when full sync finished we will not see this.

This is true and I was thinking the same. However, I don't see how else we could do something? We a) need a separate counter other than `journal_rec_executed` because mainly of multi command transactions and because the counter is interleaved so it's flaky and can't be relied on b) and because of (a) we fallback to this subpar solution of using this new variable and increment it after `tx_reader.NextTxData`. What else can we do?

> Also regarding my suggestion above it is not a bug, the only think is that we will write a ping to journal when sometimes it is not needed i.e when snapshotting from SAVE command

Mixing irrelevant flows is orthogonal and has side effects (and probably a bug at somepoint). Sorry, when I said bug, I meant it's a source of confusion and potential problems.

Reply:
journal_rec_executed and tx_reader.NextTxData are on the replica side and I dont see how they are relevant here. Inorder to make sure we get all the journal changes in replica side and that we are in sync with lsn we must send the lsn from master. Because the lsn is saved in the journal_slice we need to write the lsn value in hold by this class.
I do think that changing journal slice write entry Op::PING with the lsn is the right way. you can ignore the Op::PING in snapshoting as we do for Op::EXEC and for Op::NOOP. I dont see this as mixing irrelevant flows. We want to be able to track lsn so we must record it. Weather we want to track it is the callback which is writing the sync to deside, in snapshot no in steamer yes.
This will also simplify the flow in replica cause you will always increase the lsn on PING on master side you will always increase journal_rec_executed on replica side

Reply:
I made the changes but now we trigger the `DCHECK` almost every time on `test_replication_all` which IMO I think it should not happen. So there is either something that we are missing or a bug.

> This will also simplify the flow in replica cause you will always increase the lsn on PING on master side you will always increase journal_rec_executed on replica side

We still need a separate variable for this because `journal_rec_executed` gets incremented interleaved. See also your comment above:

```
I wrote in the issue that we need to compare with journal_rec_executed_, but actually now after reviewing the code I understand that we need a different counter to compare to which will be incremented inside NextTxData when reading another entry from socket.
The reason for this is that under use_multi_shard_exe_sync_ we accumulate the multi transaction data and do not execute it therefor you might get different value in journal_rec_executed_ when comparing when getting a ping if it is in the middle of multi transaction
```

Reply:
The changes you did are again sending pings from the streamer and not from journal slice.. this was my first comment in this thread that this will not work because when having several replicas you send the ping to only one and increase the lag for all

Reply:
I see. I made the changes, however the DCHECK still triggers. I will investigate. Let me know if you got any ideas as well